packages/cache/src/hashFile.ts

```diff
@@ -0,0 +1,52 @@
+import * as path from "path";
+import { promises as fs } from "fs";
+import * as crypto from "crypto";
+import pLimit from "p-limit";
+
+let MAX_FILE_OPERATIONS = 5000;
+
+try {
+  const maxFileOpEnv = process.env["BACKFILL_MAX_FILE_OP"];
+  if (maxFileOpEnv) {
+    MAX_FILE_OPERATIONS = parseInt(maxFileOpEnv);
+  }
+} catch (_) {
+  /* The env variable is not an integer, this is fine.*/
+}
+
+const diskLimit = pLimit(MAX_FILE_OPERATIONS);
+
+// The first key is the file path, the second key is mtime
+const memo = new Map<string, Map<number, string>>();
+
+async function computeHash(filePath: string): Promise<string> {
+  const fileBuffer = await diskLimit(() => {
+    return fs.readFile(filePath);
+  });
+  // We use sha1 for perf reason and because the hashing is not used for security reason.
+  const hashSum = crypto.createHash("sha1");
+  hashSum.update(fileBuffer);
+  const hash = hashSum.digest("hex");
+  return hash;
+}
+
+/*
+ * Get the hash of a file.
+ * This function memoizes the hash for files and mtimes.
+ */
+export async function getFileHash(
+  cwd: string,
+  filePath: string
+): Promise<string> {
+  const fileAbsPath = path.join(cwd, filePath);
+  const stat = await fs.stat(fileAbsPath);
+  if (memo.get(fileAbsPath) && memo.get(fileAbsPath)!.get(stat.mtimeMs)) {
+    return memo.get(fileAbsPath)!.get(stat.mtimeMs)!;
+  }
+  const hash = await computeHash(fileAbsPath);
+  if (!memo.get(fileAbsPath)) {
+    memo.set(fileAbsPath, new Map<number, string>());
+  }
+  memo.get(fileAbsPath)!.set(stat.mtimeMs, hash);
```

Comment:
Is there a significant perf or memory impact from using an object with `{ fileAbsPath: string, mtimeMS: number }` as key of the memo Map ?

This would also simplify the hoops here:
```typescript
  const fileAbsPath = path.join(cwd, filePath);
  const { mtimeMs } = await fs.stat(fileAbsPath);
  const key = { fileAbsPath, mtimeMs };
  let hash;
  if (!memo.has(key)) {
    hash = await computeHash(fileAbsPath);
    memo.set(key, hash);
  } else {
    hash = memo.get(key);
  }
  
  return hash;
}
```

Reply:
I am not sure I understand. Map does not do deep equality, so your cache key will never be a hit.